File: backend/go/internal/app/runtime_lease.go
package app

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"regexp"
	"time"
)

const defaultRuntimeLeaseTTL = 30 * time.Second

var runtimeLeaseEnvironmentPattern = regexp.MustCompile(`^[a-z0-9][a-z0-9_-]{0,31}$`)

type runtimeLeaseStore interface {
	TryAcquireRuntimeLease(context.Context, string, string, time.Duration) (bool, error)
	RefreshRuntimeLease(context.Context, string, string, time.Duration) (bool, error)
	ReleaseRuntimeLease(context.Context, string, string) error
}

type runtimeLeaseRecord struct {
	SchemaVersion int    `json:"schemaVersion"`
	Owner         string `json:"owner"`
	Instance      string `json:"instance"`
	AcquiredAt    int64  `json:"acquiredAt"`
}

func runtimeLeaseKey(environment string, actor string) (string, error) {
	if !runtimeLeaseEnvironmentPattern.MatchString(environment) {
		return "", errors.New("environment has an invalid shared-state namespace")
	}
	if actor != "pipeline-consumer" && actor != "backup-scheduler" {
		return "", fmt.Errorf("unsupported runtime lease actor %q", actor)
	}
	return fmt.Sprintf("cf:v1:%s:lease:%s", environment, actor), nil
}

func runtimeLeaseValue(owner string, instance string, acquiredAt time.Time) (string, error) {
	if owner != "node" && owner != "go" {
		return "", fmt.Errorf("unsupported runtime lease owner %q", owner)
	}
	if instance == "" {
		instance = randomRuntimeLeaseInstance(owner)
	}
	value, err := json.Marshal(runtimeLeaseRecord{
		SchemaVersion: 1,
		Owner:         owner,
		Instance:      instance,
		AcquiredAt:    acquiredAt.UnixMilli(),
	})
	if err != nil {
		return "", err
	}
	return string(value), nil
}

func randomRuntimeLeaseInstance(owner string) string {
	hostname, err := os.Hostname()
	if err != nil || hostname == "" {
		hostname = "unknown-host"
	}
	random := make([]byte, 16)
	if _, err := rand.Read(random); err != nil {
		return fmt.Sprintf("%s:%s:%d:%d", owner, hostname, os.Getpid(), time.Now().UnixNano())
	}
	return fmt.Sprintf("%s:%s:%d:%s", owner, hostname, os.Getpid(), hex.EncodeToString(random))
}

File: backend/nodejs/utils/runtime-lease.ts
import { hostname } from 'node:os'
import { randomUUID } from 'node:crypto'
import { assertSharedStateEnvironment } from './shared-state-contract'
import { sharedStateEnvironment } from './shared-session'
import { useSharedRedis } from './shared-redis'

export const DEFAULT_RUNTIME_LEASE_TTL_MS = 30_000

export type RuntimeLeaseActor = 'pipeline-consumer' | 'backup-scheduler'
export type RuntimeLeaseOwner = 'node' | 'go'

const RUNTIME_LEASE_ACTORS = new Set<RuntimeLeaseActor>([
  'pipeline-consumer',
  'backup-scheduler',
])
const RUNTIME_LEASE_OWNERS = new Set<RuntimeLeaseOwner>(['node', 'go'])

export interface RuntimeLeaseHandle {
  key: string
  value: string
  acquired: boolean
  enabled: boolean
  release(): Promise<void>
}

interface RuntimeLeaseOptions {
  actor: RuntimeLeaseActor
  owner: RuntimeLeaseOwner
  ttlMs?: number
  onLost?: (reason: unknown) => void | Promise<void>
}

const REFRESH_RUNTIME_LEASE_SCRIPT = `
if redis.call('GET', KEYS[1]) == ARGV[1] then
  return redis.call('PEXPIRE', KEYS[1], ARGV[2])
end
return 0
`

const RELEASE_RUNTIME_LEASE_SCRIPT = `
if redis.call('GET', KEYS[1]) == ARGV[1] then
  return redis.call('DEL', KEYS[1])
end
return 0
`

export function runtimeLeaseKey(
  actor: RuntimeLeaseActor,
  environment = sharedStateEnvironment(),
) {
  assertRuntimeLeaseActor(actor)
  return `cf:v1:${assertSharedStateEnvironment(environment)}:lease:${actor}`
}

export function runtimeLeaseValue(
  owner: RuntimeLeaseOwner,
  instance = `${hostname()}:${process.pid}:${randomUUID()}`,
) {
  assertRuntimeLeaseOwner(owner)
  return JSON.stringify({
    schemaVersion: 1,
    owner,
    instance,
    acquiredAt: Date.now(),
  })
}

function assertRuntimeLeaseActor(actor: string): asserts actor is RuntimeLeaseActor {
  if (!RUNTIME_LEASE_ACTORS.has(actor as RuntimeLeaseActor)) {
    throw new Error(`Unsupported runtime lease actor: ${actor}`)
  }
}

function assertRuntimeLeaseOwner(owner: string): asserts owner is RuntimeLeaseOwner {
  if (!RUNTIME_LEASE_OWNERS.has(owner as RuntimeLeaseOwner)) {
    throw new Error(`Unsupported runtime lease owner: ${owner}`)
  }
}

export async function acquireRuntimeLease({
  actor,
  owner,
  ttlMs = DEFAULT_RUNTIME_LEASE_TTL_MS,
  onLost,
}: RuntimeLeaseOptions): Promise<RuntimeLeaseHandle> {
  const key = runtimeLeaseKey(actor)
  const value = runtimeLeaseValue(owner)
  const redis = await useSharedRedis()
  if (!redis) {
    return {
      key,
      value,
      acquired: true,
      enabled: false,
      async release() {},
    }
  }

  const acquired = await redis.set(key, value, {
    NX: true,
    PX: ttlMs,
  })

  if (acquired !== 'OK') {
    return {
      key,
      value,
      acquired: false,
      enabled: true,
      async release() {},
    }
  }

  let closed = false
  const refreshInterval = setInterval(async () => {
    if (closed) return
    try {
      const refreshed = await redis.eval(REFRESH_RUNTIME_LEASE_SCRIPT, {
        keys: [key],
        arguments: [value, String(ttlMs)],
      })
      if (refreshed !== 1 && refreshed !== true) {
        closed = true
        clearInterval(refreshInterval)
        await onLost?.(new Error(`Runtime lease ${key} was lost`))
      }
    } catch (error) {
      closed = true
      clearInterval(refreshInterval)
      await onLost?.(error)
    }
  }, Math.max(1_000, Math.floor(ttlMs / 3)))
  refreshInterval.unref?.()

  return {
    key,
    value,
    acquired: true,
    enabled: true,
    async release() {
      if (closed) return
      closed = true
      clearInterval(refreshInterval)
      await redis.eval(RELEASE_RUNTIME_LEASE_SCRIPT, {
        keys: [key],
        arguments: [value],
      })
    },
  }
}
